### PR TITLE
Reproject-core

### PR DESCRIPTION
**Pull request by aquatiko**
I've added the core logic behind the reproject.
- [X] Function
- [X] Test
- [X] Doc

### CODE REVIEW

File: src/Reproject.jl [hunk 1]
```diff
@@ -1,8 +1,14 @@
...
 
 include("parsers.jl")
 include("utils.jl")
+include("core.jl")
+
+export
+    reproject
 
 end # module
+
```
Comment (giordano) [inline src/Reproject.jl]: Again: you don't need to add an extra empty line, just make sure that the last line is complete, i.e., it's ended by a newline character

File: src/core.jl [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        end
+    end
+    
+    if return_footprint
+        return img_out, map(x -> x !== NaN ? 1.0 : 0.0, img_out)
+    else
+        return img_out
+    end
+end
+
```
Comment (giordano) [inline src/core.jl]: Also here: no need for an extra empty line at the end

File: test/runtests.jl [hunk 3]
```diff
@@ -3,12 +3,14 @@ using Test
 using Conda, PyCall
 using FITSIO, WCS
 
-ENV["PYTHON"]=""
+#ENV["PYTHON"]=""
```
Comment (giordano) [inline test/runtests.jl]: Why?

File: src/core.jl [hunk 4]
```diff
@@ -0,0 +1,70 @@
+function reproject(input_data, output_projection; shape_out = nothing, order = 1, hdu_in = 1, hdu_out = 1, return_footprint = true)
```
Comment (giordano) [inline src/core.jl]: I don't want the `return_footprint` argument, as it makes the function type unstable.  Remove it and always return the footprint (as explained below).  This is one of the Julia performance tips that I told you to keep in mind.  Please, read them again.

In general, always make sure that the function is type stable with `@code_warntype`.  I haven't run but I'm already sure that you didn't :wink:.

File: src/core.jl [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            pix_cord_out = world_to_pix(wcs_out, [rad2deg(lon(cord_out)), rad2deg(lat(cord_out))])
+
+            if 1 <= pix_cord_out[1] <= size(array_in)[1] && 1 <= pix_cord_out[2] <= size(array_in)[2] 
+                img_out[i,j] = itp(pix_cord_out[1], pix_cord_out[2])
+            end
+        end
+    end
+    
+    if return_footprint
+        return img_out, map(x -> x !== NaN ? 1.0 : 0.0, img_out)
```
Comment (giordano) [inline src/core.jl]: The footprint is simply `(!isnan).(img_out)`

File: src/core.jl [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    end
+    
+    type_in = wcs_to_celestial_frame(wcs_in)
+    type_out = wcs_to_celestial_frame(wcs_out)
+    
+    if type_in == type_out
+        return array_in
+    end
+    
+    img_out = fill!(zeros(shape_out), NaN)
```
Comment (giordano) [inline src/core.jl]: Are you sure it's a good idea to initialise the matrix to NaN?  Why not zeros or even `undef`?  As it is now you're setting the memory twice, which is a huge waste of computation power, as you're going to set all elements anyway later, aren't you?

File: src/core.jl [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        itp = interpolate(array_in, BSpline(Constant()))
+    elseif order == 1
+        itp = interpolate(array_in, BSpline(Linear()))
+    else
+        itp = interpolate(array_in, BSpline(Quadratic(InPlace(OnCell()))))
+    end
+    
+    for i in 1:shape_out[1]
+        for j in 1:shape_out[2]
+            pix_cord_in = [float(i); float(j)]
```
Comment (giordano) [inline src/core.jl]: Is this supposed to be called `pix_coord_in`?  Also, does it need to be a vector or it can be a tuple as well?

File: src/core.jl [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+                cord_out = convert(ICRSCoords, cord_in)
+            elseif type_out == "Gal"
+                cord_out = convert(GalCoords, cord_in)
+            elseif type_out == "FK5"
+                cord_out = convert(FK5Coords{wcs_out.equinox}, cord_in)
+            else
+                throw(ArgumentError("Unsupported output WCS coordinate type"))
+            end
+
+            pix_cord_out = world_to_pix(wcs_out, [rad2deg(lon(cord_out)), rad2deg(lat(cord_out))])
```
Comment (giordano) [inline src/core.jl]: Is this supposed to be called `pix_coord_out`?  Also, does it need to be a vector or it can be a tuple as well?

File: test/core.jl [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        diff = 10e-5
+        count = 0
+        for i in 1 : size(arr1)[1] - 1
+            for j in 1 : size(arr2)[2] - 1
+                if abs(arr1[i,j] - arr2[j,i]) > diff
+                    count+=1
+                end
+            end
+        end
+        return  count == 0 ? true : false
```
Comment (giordano) [inline test/core.jl]: Just return `count == 0` (or even better  `iszero(count)`), no?

*Edit*: however see the comment about how to replace the whole function with something way simpler

File: test/core.jl [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        download("https://astropy.stsci.edu/data/galactic_center/gc_msx_e.fits", "gc_msx_e.fits")
+    end
+    
+    imgin = FITS("gc_msx_e.fits")         # project this
+    imgout = FITS("gc_2mass_k.fits")        # into this coordinate
+    
+    hdu1 = astropy.io.fits.open("gc_2mass_k.fits")[1]
+    hdu2 = astropy.io.fits.open("gc_msx_e.fits")[1]
+    
+    function check_diff(arr1, arr2)
```
Comment (giordano) [inline test/core.jl:17]: I think the whole function is simply
```julia
all(isapprox(arr1, arr2))
```
or
```julia
all(isapprox(arr1, arr2, abstol = 1e-4))
```
if you want to specify a different absolute tolerance

File: test/core.jl [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+        diff = 10e-5
```
Comment (giordano) [inline test/core.jl]: Note that `10e-5` is `1e-4`

**Review by giordano — CHANGES_REQUESTED** [review]
I think it's overall good.  However, now it's time to have a look at performance.  These functions can be used with huge images (also several hundreds of MiB), every single CPU single can make a difference.  Make sure the functions are type stable.

File: test/core.jl [hunk 12]
```diff
@@ -0,0 +1,35 @@
+#using Reproject: wcs_to_celestial_frame
+@testset "reproject-core" begin
+    if !isfile("gc_2mass_k.fits")
```
Comment (giordano) [inline test/core.jl]: Also, please save the files to `test/data`

**Comment by giordano** [comment]
Another important factor: the output array **must** have the same type as the input one.  You need to mark the input array as `::AbstractArray{T}` and use the type `T` as appropriate.  Tests should make sure that the returned array is of the expected type (besides type stability in general)

File: src/core.jl (hunk 7 above)
Comment (aquatiko) [reply src/core.jl]: `pix_to_world` expects it to be a vector. Also, the name is intuitively correct as it it the pixel coordinates(from input image) which is getting converted to desired output image.

File: test/runtests.jl (hunk 3 above)
Comment (aquatiko) [reply test/runtests.jl]: I wasn't sure what to add here. Is there a need to specify a specific python version here?

File: src/core.jl (hunk 6 above)
Comment (aquatiko) [reply src/core.jl]: Would `fill(NaN, shape_out)` be good?
I need `NaN` filled arrays to form the footprint of output image as done in python version.
I can't take any integer filled number as base as it will not form the correct grayscaled image.

File: test/runtests.jl (hunk 3 above)
Comment (giordano) [reply test/runtests.jl]: It's not clear to me why you wanted to change this line at all.  This is to force Conda's Python.

File: src/core.jl (hunk 7 above)
Comment (giordano) [reply src/core.jl]: I was referring to the fact that probably "cord" stands for "coordinates", with two "o"s rather than one.

File: test/core.jl (hunk 10 above)
Comment (aquatiko) [reply test/core.jl:17]: I tried using this function, in my version of reproject some corner cases arises where we get NaN but some numerical value in python version. `check_diff` is written in a specific way to ignore those minor differences.

Comment (giordano) [reply test/core.jl:17]: I don't see how this function treats `NaN` specially:
```julia
julia> A = [NaN  2.4 5.9
            -3.6 NaN 0.18
            -1.7 6.0 2.3]
3×3 Array{Float64,2}:
 NaN      2.4  5.9 
  -3.6  NaN    0.18
  -1.7    6.0  2.3 

julia> check_diff(A, A)
false

julia> all(isapprox(A, A, nans = true))
true
```
`check_diff` wrongly reports that `A` and `A` are different when they're actually the same matrix.  `isapprox` with `nans=true` correctly says that they are approximately equal.

Comment (aquatiko) [reply test/core.jl:17]: `check_diff` transpose the second array as one comes from python. Thus `check_diff(A,A)` won't be true here.

**Comment by aquatiko** [comment]
> Another important factor: the output array **must** have the same type as the input one. You need to mark the input array as `::AbstractArray{T}` and use the type `T` as appropriate. Tests should make sure that the returned array is of the expected type (besides type stability in general)

I'm not taking the input data directly. I'm extracting it from a  `ImageHDU` or a `FITS` file.  I can specify the output type but where do I specify the type of input array then?
I can make the parsers return the input array in specific format but here we are only reprojecting 2D images. Is AbstractArray necessary here?

**Comment by aquatiko** [comment]
The type of things returned here is `Tuple{Array{Float64,2},BitArray{2}}`, is it fine to assign this as return type?

**Comment by giordano** [comment]
Yes, it's fine as long it's stable.

**Comment by giordano** [comment]
> I'm not taking the input data directly.

Have a function that takes as input an array: https://docs.julialang.org/en/v1/manual/performance-tips/#kernel-functions-1

**Comment by aquatiko** [comment]
Need a CI re-run!

File: src/core.jl [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    return img_out, (!isnan).(img_out)
+end
+
+
+"""
+    interpolator(array_in, order::Int)
+
+Returns an interpolator with the given array and order of interpolation.
+"""
+function interpolator(array_in::AbstractArray{T}, order::Int) where {T}
```
Comment (giordano) [inline src/core.jl]: The `T` parameter is not actually used, just do
```suggestion
function interpolator(array_in::AbstractArray, order::Int)
```

File: src/core.jl [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+                       Frame can be taken from WCSTransform object, ImageHDU, FITS or name of FITS file.
+- `shape_out`: Shape of image after reprojection.
+- `order`: Order of interpolation.
+           0: Nearest-neighbor
+           1: Linear
+           2: Quadratic
+- `hdu_in`: Used to specify HDU number when giving input as FITS or name of FITS file.
+- `hud_out:` Used to specify HDU number when giving output projection as FITS or name of FITS file.
+"""
+function reproject(input_data, output_projection; shape_out = nothing, order::Int = 1, hdu_in::Int = 1, hdu_out::Int = 1)::Tuple{Array{Float64,2},BitArray{2}}
```
Comment (giordano) [inline src/core.jl]: Is the type signature of the output really necessary for type stability?  I almost never need to use it.

**Review by giordano — COMMENTED** [review]
Only a couple of comments!

**Comment by giordano** [comment]
Also, code coverage can be probably improved.  Errors are important to be tested, too :wink:

**Comment by aquatiko** [comment]
@giordano incorporated your suggestions. Coverage Improved.
